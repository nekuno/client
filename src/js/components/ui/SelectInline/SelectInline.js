import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styles from './SelectInline.scss';

export default class SelectInline extends Component {

    static propTypes = {
        title          : PropTypes.string,
        options        : PropTypes.array.isRequired,
        defaultOption  : PropTypes.string,
        multiple       : PropTypes.bool,
        color          : PropTypes.oneOf(['purple', 'blue', 'pink', 'green']),
        nullable       : PropTypes.bool,
        onClickHandler : PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            selected: props.defaultOption ? [props.defaultOption] : []
        };
    }

    handleClick(value) {
        const {selected} = this.state;
        const {multiple, nullable} = this.props;
        const index = selected.indexOf(value);
        let newSelected = [];

        if (index !== -1 && (nullable || selected.length > 1)) {
            newSelected = selected.filter(option => option !== value);
        } else if (index !== -1 && !nullable && selected.length === 1) {
            newSelected = selected;
        } else if (multiple) {
            newSelected = [...this.state.selected, value];
        } else {
            newSelected = [value];
        }
        this.setState({selected: newSelected});

        if (this.props.onClickHandler) {
            this.props.onClickHandler(newSelected);
        }
    }

    render() {
        const {title, options, color} = this.props;
        const {selected} = this.state;
        const optionWidthPercent = 100 / options.length;

        return (
            <div>
                {title ? <div className={styles.title + ' small'}>{title}</div> : null}
                <div className={styles.selectInlineWrapper + ' ' +  styles[color]}>
                    <div className={styles.selectInline}>
                        {options.map(option => {
                            let optionClass = selected.some(selectedOption => selectedOption === option.id) ? styles.optionWrapper + ' ' + styles.selected : styles.optionWrapper;
                            optionClass = optionClass + ' ' + styles[color];
                            return (
                                <div key={option.id} className={optionClass + ' ' + styles[color]} onClick={this.handleClick.bind(this, option.id)} style={{width: optionWidthPercent + '%'}}>
                                    <div className={styles.option + ' ' + styles[color]}>
                                        <div className={styles.optionText}>
                                            {option.text}
                                        </div>
                                    </div>
                                </div>
                            )
                        })}
                    </div>
                </div>
            </div>
        );
    }
}