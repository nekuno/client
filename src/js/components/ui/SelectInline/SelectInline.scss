@import '../../../../scss/_partials/colors';

.title {
  font-weight: 600;
  margin-bottom: 16px;
}
.select-inline-wrapper {
  border: 1px solid $gray-border;
  border-radius: 18px;
  cursor: pointer;
  -moz-border-radius: 18px;
  -webkit-border-radius: 18px;
  height: 36px;
  margin: auto;
  overflow: hidden;
  text-align: center;
  width: 90%;

  .select-inline {
    border-radius: 18px;
    -moz-border-radius: 18px;
    -webkit-border-radius: 18px;
    height: 38px;

    .option-wrapper {
      background: $white-background;
      color: $gray-secondary-text;
      display: inline-block;
      height: 36px;
      line-height: 36px;
      text-align: center;

      .option {
        border-left: 1px solid $gray-border;
        border-right: 1px solid $gray-border;
        width: 100%;

        .option-text {
          margin: auto;
          white-space: nowrap;
          overflow: hidden;
          text-align: center;
          text-overflow: ellipsis;
          width: 90%;
        }
      }

      &:first-of-type {
        .option {
          border-left: none;
        }
      }

      &.selected {
        background: $purple-main;
        color: $white-background;

        &.purple {
          background: $purple-main;
        }
        &.blue {
          background: $blue-projects;
        }
        &.pink {
          background: $pink-hobbies;
        }
        &.green {
          background: $green-experiences;
        }
      }
    }
  }
}