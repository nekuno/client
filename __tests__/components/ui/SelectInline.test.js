import React from 'react';
import SelectInline from '../../../src/js/components/ui/SelectInline/SelectInline.js';
import styles from '../../../src/js/components/ui/SelectInline/SelectInline.scss';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


const options = [
    {
        id: "lorem",
        text: "Lorem"
    },
    {
        id: "ipsum",
        text: "Ipsum"
    },
    {
        id: "lorem-ipsum",
        text: "Lorem ipsum"
    },
    {
        id: "sed-ut-perspiciatis",
        text: "Sed ut perspiciatis"
    }
];

describe('Test SelectInline component', () => {
    const mockFn = jest.fn();
    const selectInline = shallow(<SelectInline options={options} onClickHandler={mockFn}/>);

    it('should be defined', () => {
        expect(SelectInline).toBeDefined();
    });
    it('should render correctly', () => {
        expect(selectInline).toMatchSnapshot();
    });
    it('SelectInline click event', () => {
        selectInline.find(`.${styles.optionWrapper}`).first().simulate('click');
        expect(mockFn).toHaveBeenCalledTimes(1);
    });
});
