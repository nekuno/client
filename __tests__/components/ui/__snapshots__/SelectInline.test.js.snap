// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SelectInline component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectInline
    onClickHandler={[MockFunction]}
    options={
      Array [
        Object {
          "id": "lorem",
          "text": "Lorem",
        },
        Object {
          "id": "ipsum",
          "text": "Ipsum",
        },
        Object {
          "id": "lorem-ipsum",
          "text": "Lorem ipsum",
        },
        Object {
          "id": "sed-ut-perspiciatis",
          "text": "Sed ut perspiciatis",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <div
          className="selectInlineWrapper"
        >
          <div
            className="selectInline"
          >
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Lorem
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Ipsum
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Lorem ipsum
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Sed ut perspiciatis
                </div>
              </div>
            </div>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="selectInline"
          >
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Lorem
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Ipsum
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Lorem ipsum
                </div>
              </div>
            </div>
            <div
              className="optionWrapper undefined"
              onClick={[Function]}
              style={
                Object {
                  "width": "25%",
                }
              }
            >
              <div
                className="option"
              >
                <div
                  className="optionText"
                >
                  Sed ut perspiciatis
                </div>
              </div>
            </div>
          </div>,
          "className": "selectInlineWrapper",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem
                  </div>
                </div>
              </div>,
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Ipsum
                  </div>
                </div>
              </div>,
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem ipsum
                  </div>
                </div>
              </div>,
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Sed ut perspiciatis
                  </div>
                </div>
              </div>,
            ],
            "className": "selectInline",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "lorem",
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem
                  </div>
                </div>,
                "className": "optionWrapper undefined",
                "onClick": [Function],
                "style": Object {
                  "width": "25%",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="optionText"
                  >
                    Lorem
                  </div>,
                  "className": "option",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Lorem",
                    "className": "optionText",
                  },
                  "ref": null,
                  "rendered": "Lorem",
                  "type": "div",
                },
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "ipsum",
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Ipsum
                  </div>
                </div>,
                "className": "optionWrapper undefined",
                "onClick": [Function],
                "style": Object {
                  "width": "25%",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="optionText"
                  >
                    Ipsum
                  </div>,
                  "className": "option",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Ipsum",
                    "className": "optionText",
                  },
                  "ref": null,
                  "rendered": "Ipsum",
                  "type": "div",
                },
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "lorem-ipsum",
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem ipsum
                  </div>
                </div>,
                "className": "optionWrapper undefined",
                "onClick": [Function],
                "style": Object {
                  "width": "25%",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="optionText"
                  >
                    Lorem ipsum
                  </div>,
                  "className": "option",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Lorem ipsum",
                    "className": "optionText",
                  },
                  "ref": null,
                  "rendered": "Lorem ipsum",
                  "type": "div",
                },
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "sed-ut-perspiciatis",
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Sed ut perspiciatis
                  </div>
                </div>,
                "className": "optionWrapper undefined",
                "onClick": [Function],
                "style": Object {
                  "width": "25%",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="optionText"
                  >
                    Sed ut perspiciatis
                  </div>,
                  "className": "option",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Sed ut perspiciatis",
                    "className": "optionText",
                  },
                  "ref": null,
                  "rendered": "Sed ut perspiciatis",
                  "type": "div",
                },
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <div
            className="selectInlineWrapper"
          >
            <div
              className="selectInline"
            >
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Ipsum
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem ipsum
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Sed ut perspiciatis
                  </div>
                </div>
              </div>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="selectInline"
            >
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Ipsum
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Lorem ipsum
                  </div>
                </div>
              </div>
              <div
                className="optionWrapper undefined"
                onClick={[Function]}
                style={
                  Object {
                    "width": "25%",
                  }
                }
              >
                <div
                  className="option"
                >
                  <div
                    className="optionText"
                  >
                    Sed ut perspiciatis
                  </div>
                </div>
              </div>
            </div>,
            "className": "selectInlineWrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="optionWrapper undefined"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Lorem
                    </div>
                  </div>
                </div>,
                <div
                  className="optionWrapper undefined"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Ipsum
                    </div>
                  </div>
                </div>,
                <div
                  className="optionWrapper undefined"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Lorem ipsum
                    </div>
                  </div>
                </div>,
                <div
                  className="optionWrapper undefined"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": "25%",
                    }
                  }
                >
                  <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Sed ut perspiciatis
                    </div>
                  </div>
                </div>,
              ],
              "className": "selectInline",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "lorem",
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Lorem
                    </div>
                  </div>,
                  "className": "optionWrapper undefined",
                  "onClick": [Function],
                  "style": Object {
                    "width": "25%",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <div
                      className="optionText"
                    >
                      Lorem
                    </div>,
                    "className": "option",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Lorem",
                      "className": "optionText",
                    },
                    "ref": null,
                    "rendered": "Lorem",
                    "type": "div",
                  },
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": "ipsum",
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Ipsum
                    </div>
                  </div>,
                  "className": "optionWrapper undefined",
                  "onClick": [Function],
                  "style": Object {
                    "width": "25%",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <div
                      className="optionText"
                    >
                      Ipsum
                    </div>,
                    "className": "option",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Ipsum",
                      "className": "optionText",
                    },
                    "ref": null,
                    "rendered": "Ipsum",
                    "type": "div",
                  },
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": "lorem-ipsum",
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Lorem ipsum
                    </div>
                  </div>,
                  "className": "optionWrapper undefined",
                  "onClick": [Function],
                  "style": Object {
                    "width": "25%",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <div
                      className="optionText"
                    >
                      Lorem ipsum
                    </div>,
                    "className": "option",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Lorem ipsum",
                      "className": "optionText",
                    },
                    "ref": null,
                    "rendered": "Lorem ipsum",
                    "type": "div",
                  },
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": "sed-ut-perspiciatis",
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    className="option"
                  >
                    <div
                      className="optionText"
                    >
                      Sed ut perspiciatis
                    </div>
                  </div>,
                  "className": "optionWrapper undefined",
                  "onClick": [Function],
                  "style": Object {
                    "width": "25%",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <div
                      className="optionText"
                    >
                      Sed ut perspiciatis
                    </div>,
                    "className": "option",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Sed ut perspiciatis",
                      "className": "optionText",
                    },
                    "ref": null,
                    "rendered": "Sed ut perspiciatis",
                    "type": "div",
                  },
                  "type": "div",
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
